src/solvers/refinement/string_refinement.cpp

```diff
@@ -311,33 +311,81 @@ bool is_char_pointer_type(const typet &type)
 }
 
 /// \param type: a type
-/// \param ns: name space
-/// \return true if a subtype is an pointer of characters
-static bool has_char_pointer_subtype(const typet &type, const namespacet &ns)
+/// \param pred: a predicate
+/// \return true if one of the subtype of `type` satisfies predicate `pred`.
+///         The meaning of "subtype" is in the algebraic datatype sense:
+///         for example, the subtypes of a struct are the types of its
+///         components, the subtype of a pointer is the type it points to,
+///         etc...
+///         For instance in the type `t` defined by
+///         `{ int a; char[] b; double * c; { bool d} e}`, `int`, `char`,
+///         `double` and `bool` are subtypes of `t`.
+bool has_subtype(
+  const typet &type, const std::function<bool(const typet &)> &pred)
 {
-  if(is_char_pointer_type(type))
+  if(pred(type))
     return true;
 
+  // clang-format off
   if(type.id() == ID_struct || type.id() == ID_union)
   {
     const struct_union_typet &struct_type = to_struct_union_type(type);
-    for(const auto &comp : struct_type.components())
-    {
-      if(has_char_pointer_subtype(comp.type(), ns))
-        return true;
-    }
+    return std::any_of(
+      struct_type.components().begin(),
+      struct_type.components().end(),
+      [&](const struct_union_typet::componentt &comp)
+      {
+        return has_subtype(comp.type(), pred);
+      });
   }
 
-  for(const auto &t : type.subtypes())
-  {
-    if(has_char_pointer_subtype(t, ns))
-      return true;
-  }
-  return false;
+  return std::any_of(
+    type.subtypes().begin(), type.subtypes().end(), [&](const typet &t)
+    {
+      return has_subtype(t, pred);
+    });
+  // clang-format on
+}
+
+/// \param type: a type
+/// \param ns: namespace
+/// \return true if a subtype of `type` is an pointer of characters.
+///         The meaning of "subtype" is in the algebraic datatype sense:
+///         for example, the subtypes of a struct are the types of its
+///         components, the subtype of a pointer is the type it points to,
+///         etc...
+static bool has_char_pointer_subtype(const typet &type)
+{
+  // clang-format off
+  return has_subtype(
+    type,
+    [](const typet &subtype)
```

Comment:
Think you can just pass in `is_char_pointer_type` rather than wrapping in predicate

```C++
return has_subtype(type, is_char_pointer_type);
```